### ISSUE TITLE
Test.afterTest is not called with Configuration Cache

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#25311]
### Expected Behavior

```
configurationCache-afterTest_missing>gradlew test --rerun-tasks --configuration-cache
Calculating task graph as no configuration cache is available for tasks: test

> Task :test
com.example.MyTest > myTest: SUCCESS

BUILD SUCCESSFUL in 3s
2 actionable tasks: 2 executed
Configuration cache entry stored.
```

### Current Behavior

```
configurationCache-afterTest_missing>gradlew test --rerun-tasks --configuration-cache
Calculating task graph as no configuration cache is available for tasks: test

BUILD SUCCESSFUL in 3s
2 actionable tasks: 2 executed
Configuration cache entry stored.
```

### Context (optional)

When I run the same thing without configuration cache:
```
configurationCache-afterTest_missing>gradlew test --rerun-tasks
> Task :test
com.example.MyTest > myTest: SUCCESS

BUILD SUCCESSFUL in 3s
2 actionable tasks: 2 executed
```

Originally I had `logger.quiet(...)` instead of `println()`, changed it only to make sure there's no captured Project / Task object in play.

### Steps to Reproduce

1. Clone https://github.com/TWiStErRob/repros/tree/main/gradle/configurationCache-afterTest_missing
2. `gradlew test --rerun-tasks` (I ran it twice to ensure it always executes)
3. `gradlew test --rerun-tasks --configuration-cache`

(Note `--rerun-tasks` is not necessary for the repro if you delete the `build` and `.gradle` folders between executions.)

### Gradle version

repro 8.2-rc-1, tried also 8.0.2

### Your Environment (optional)

I reproduced this on Windows 10 / Java 17. Discovered it inside a Gradle Test Kit test initially, then separated out a minimal repro.

**Comment by bamboo** [comment]
Related to #24613

**Comment by abstratt** [comment]
@TWiStErRob Could you please confirm whether you still see this behavior with tonight's nightly? (builds at midnight)

https://services.gradle.org/distributions-snapshots/

**Comment by TWiStErRob** [comment]
Still reproduces with https://services.gradle.org/distributions-snapshots/gradle-8.3-20230613222630+0000-all.zip

Updated the repro, because I committed the wrong version, sorry about the confusion.

**Comment by abstratt** [comment]
@TWiStErRob This time I think we fixed this, when addressing #24613 and friends.

Could you please try with a recent 8.4 build? I saw this working fine afterI ran:

```
./gradlew wrapper --gradle-version=8.4-20230828223511+0000
```

**Comment by TWiStErRob** [comment]
Based on my repro in OP, it is indeed fixed in the quoted version of yesterday's nightly.

**Comment by TWiStErRob** [comment]
Do you know what fixed it? because the linked #24613's fix only seems to modify TestNG related files.

**Comment by abstratt** [comment]
But actually answering your question, I believe the fix was related to simplifying how the`AbstractTestTask` state for managing listeners was serialized (#25983).

**Comment by TWiStErRob** [comment]
Thank you!

**Comment by abstratt** [comment]
Fixed by #25983.
